File: Exercise_3.java

import java.io.*;

// Java program to implement
// a Singly Linked List
public class LinkedList {

    Node head, tail; // head of list
    int length ;

    // Linked list Node.
    // This inner class is made static
    // so that main() can access it
    static class Node {

        int data;
        Node next;

        // Constructor
        Node(int d)
        {
            //Write your code here
            this.data = d;
            this.next = null;
        }
    }

    public LinkedList()
    {
        this.head = null;
        this.tail = null;
        this.length= 0;
    }

    // Method to insert a new node
    public static LinkedList insert(LinkedList list, int data)
    {
        // Create a new node with given data
        // If the Linked List is empty,
        // then make the new node as head
        // Else traverse till the last node
        // and insert the new_node there
        // Insert the new_node at last node
        // Return the list by head

        Node temp_node = new Node(data);
        if(list.head == null)
        {
            list.head = temp_node;
            list.tail = temp_node;
            list.length++;
        }
        else
        {
            list.tail.next = temp_node;
            list.tail = temp_node;
            list.length++;
        }
        return list;
    }

    // Method to print the LinkedList.
    public static void printList(LinkedList list)
    {
//         Traverse through the LinkedList
//         Print the data at current node
//         Go to next node
        Node current_node = list.head;
        while (list.length  > 0)
        {
            System.out.println(" "+current_node.data);
            current_node = current_node.next;
            list.length--;
        }
    }

    // Driver code
    public static void main(String[] args)
    {
        /* Start with the empty list. */
        LinkedList list = new LinkedList();

        //
        // ******INSERTION******
        //

        // Insert the values
        list = insert(list, 1);
        list = insert(list, 2);
        list = insert(list, 3);
        list = insert(list, 4);
        list = insert(list, 5);

        // Print the LinkedList
        printList(list);
    }
}

// Time Complexity : 
//Insert : O(1)
//Remove: O(n)
//Print: O(n)
// Space Complexity :
// Did this code successfully run on Leetcode : Didn't find any relevant question on Leetcode. 
// Any problem you faced while coding this : Pointing to and from the head and other nodes. 


// Your code here along with comments explaining your approach

/*
I am maintaining a Head and a Tail and continuously updating Tail after every insertion 
I am also maintaining a Length variable to get the length of Linkedlist. 
and Based on length, I am traversing into the lIkedlist and printing the Respective Node data. 
*/

//Reference: https://www.geeksforgeeks.org/implementing-a-linked-list-in-java-using-class/

File: Exercise_3.py
class ListNode:
    """
    A node in a singly-linked list.
    """

    def __init__(self, data=None, next=None):
        self.data = data
        self.next = next


class SinglyLinkedList:
    def __init__(self):
        """
        Create a new singly-linked list.
        Takes O(1) time.
        """
        self.head = None

    def append(self, data):
        """
        Insert a new element at the end of the list.
        Takes O(n) time.
        """
        if self.head is None:
            self.head = ListNode(data, None)
        else:
            itr = self.head
            while itr:
                if itr.next is None:
                    itr.next = ListNode(data, None)
                    return
                itr = itr.next

    def find(self, key):
        """
        Search for the first element with `data` matching
        `key`. Return the element or `None` if not found.
        Takes O(n) time.
        """
        if self.head is None:
            return
        itr = self.head
        while itr:
            if itr.data == key:
                return itr.data
            itr = itr.next
        return None

    def remove(self, key):
        """
        Remove the first occurrence of `key` in the list.
        Takes O(n) time.
        """
        if self.head is None:
            print('Cannot remove from empty list!')
            return
        if self.head.data == key:
            self.head = self.head.next
            return

        itr = self.head
        prev = ListNode()
        while itr:
            curr = itr
            if itr.data == key:
                prev.next = curr.next
                return
            prev = curr
            itr = itr.next

    def print_sll(self):
        if self.head is None:
            print('Empty Singly Linked List!')
            return

        itr = self.head
        linkedlist = ''
        while itr:
            linkedlist += str(itr.data) + '-->'
            itr = itr.next
        linkedlist += 'None'
        print(linkedlist)


if __name__ == '__main__':
    list_nums = []
    sll = SinglyLinkedList()
    sll.append(1)
    sll.append(2)
    sll.append(3)
    sll.append(4)
    sll.append(5)
    sll.find(5)
    sll.remove(2)
    sll.remove(5)
    sll.remove(1)
    sll.print_sll()
